src/task_controller.js

```diff
@@ -0,0 +1,153 @@
+   * @param {object} db - object of Database interface
+   * */
+  constructor(db) {
+    this.db = db;
+  }
+
+  /**
+   * Retrieves all tasks
+   * @return {array}
+   */
+  getAll() {
+    return this.db.getAll(type).map((e) => toTask(e));
+  }
+
+  /**
+   * Retrieves all undone tasks in order of deadline closeness
+   * @return {array}
+   */
+  getDue() {
+    const tasks = this.db.getAll(type).map((e) => toTask(e));
+    const due = tasks.filter((e) => e.completed === undefined);
+    sortByDate(due);
+    return due;
+  }
+
+  /**
+   * Completes a task and returns updated ID
+   * @param {string | number} id
+   * @return {string | number}
+   */
+  complete(id) {
+    const task = this.db.getAll(type).find((e) => e.id === id);
+    if (task === undefined) {
+      throw Error(`No task with ${id} exists`);
+    }
+    if (task.completed) {
+      throw Error('The task is already completed');
+    }
+    const now = new Date();
+    const changes = {'id': task.id, 'completed': now.getTime()};
+    return this.db.edit(type, changes);
+  }
+
+  /**
+   * Appends a task to database and returns new task ID
+   * @param {object} task
+   * @return {string | number}
+   */
+  add(task) {
+    const entity = toEntity(task);
+    if (entity.completed !== undefined) {
+      throw Error('Cannot add an already completed task');
+    }
+    return this.db.add(type, entity);
+  }
+
+  /**
+   * Edits a task's content, but not completion status
+   * @param {object} changes
+   * @return {string | number}
+   */
+  edit(changes) {
+    const entity = toEntity(changes);
+    if (entity.completed !== undefined) {
+      throw Error('Cannot edit completion time or status');
+    }
+    return this.db.edit(type, entity);
+  }
+
+  /**
+   * Retrieves all undone tasks with missed deadlines and sorts them
+   * @return {array}
+   */
+  getOverdue() {
+    const tasks = this.db.getAll(type).map((e) => toTask(e));
+    const now = new Date().getTime();
+    const overdue = tasks.filter((e) => {
+      return e.completed === undefined &&
+        e.deadline !== undefined &&
+        e.deadline < now;
+    });
+    sortByDate(overdue);
+    return overdue;
+  }
+
+  /**
+   * Deletes a task with specified ID
+   * @param {string | number} id
+   */
+  delete(id) {
+    this.db.delete(type, id);
+  }
+
+  /**
+   * Retrieves all completed tasks in order of their completion
+   * @return {array}
+   */
+  getCompleted() {
+    const tasks = this.db.getAll(type).map((e) => toTask(e));
+    const res = tasks.filter((e) => e.completed !== undefined);
+    sortByDate(res, 'completed', -1);
+    return res;
+  }
+}
+
+const taskKeys = ['id', 'title', 'description', 'deadline', 'completed'];
+
+const toEntity = (task) => {
+  if (!task.title) {
+    throw Error('Task should contain a title');
+  }
+  const entity = {};
+  for (const key in task) {
+    if (!taskKeys.includes(key)) continue;
+    entity[key] = task[key];
+  }
+  return entity;
```

Comment:
Here you have almost the same code as in `toTask`. I think it would be better to take out it into a separate function.

Reply:
Agreed, this should be generalized, thanks!